go/obscuronode/enclave/sql/edgelessdb.go

```diff
@@ -0,0 +1,523 @@
+		SerialNumber:          big.NewInt(1),
+		IsCA:                  true,
+		BasicConstraintsValid: true,
+		// this subject must match the subject authorised in the manifest.json
+		Subject:   pkix.Name{CommonName: certIssuer},
+		NotBefore: time.Now(),
+		NotAfter:  time.Now().AddDate(10, 0, 0),
+		DNSNames:  []string{enclaveHostName},
+	}
+	caPrivKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return "", "", "", fmt.Errorf("failed to generate key for CA cert to init Edgeless DB - %w", err)
+	}
+	caBytes, err := x509.CreateCertificate(rand.Reader, caCert, caCert, &caPrivKey.PublicKey, caPrivKey)
+	if err != nil {
+		return "", "", "", fmt.Errorf("failed to create CA cert - %w", err)
+	}
+	caPEMBuf := new(bytes.Buffer)
+	err = pem.Encode(caPEMBuf, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: caBytes,
+	})
+	if err != nil {
+		return "", "", "", fmt.Errorf("failed to create ca cert pem - %w", err)
+	}
+
+	userCert := &x509.Certificate{
+		SerialNumber: big.NewInt(2),
+		// the issuer and subject have to match those submitted in manifest.json
+		Issuer:    pkix.Name{CommonName: certIssuer},
+		Subject:   pkix.Name{CommonName: certSubject},
+		NotBefore: time.Now(),
+		NotAfter:  time.Now().AddDate(10, 0, 0),
+	}
+	certPrivKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return "", "", "", fmt.Errorf("failed to generate private key for user cert - %w", err)
+	}
+	userCertBytes, err := x509.CreateCertificate(rand.Reader, userCert, caCert, &certPrivKey.PublicKey, caPrivKey)
+	if err != nil {
+		return "", "", "", fmt.Errorf("failed to prepare user certificate - %w", err)
+	}
+
+	userCertPEM := new(bytes.Buffer)
+	err = pem.Encode(userCertPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: userCertBytes,
+	})
+	if err != nil {
+		return "", "", "", fmt.Errorf("failed to PEM encode user certificate - %w", err)
+	}
+
+	certKeyPEM := new(bytes.Buffer)
+	privKeyOut, err := x509.MarshalPKCS8PrivateKey(certPrivKey)
+	if err != nil {
+		return "", "", "", fmt.Errorf("failed to marshal cert priv key - %w", err)
+	}
+	err = pem.Encode(certKeyPEM, &pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: privKeyOut,
+	})
+	if err != nil {
+		return "", "", "", fmt.Errorf("failed to pem encode the user private key - %w", err)
+	}
+
+	return caPEMBuf.String(), userCertPEM.String(), certKeyPEM.String(), nil
+}
+
+func prepareEDBHTTPClient(edbPEM string) (*http.Client, error) {
+	caCertPool := x509.NewCertPool()
+	if ok := caCertPool.AppendCertsFromPEM([]byte(edbPEM)); !ok {
+		return nil, fmt.Errorf("failed to append to CA cert from edb cert pem")
+	}
+
+	return &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs:    caCertPool,
+				MinVersion: tls.VersionTLS12,
+			},
+		},
+	}, nil
+}
+
+// perform the SGX enclave attestation to verify edb running in a legit enclave and with expected edb version etc.
+func performEDBRemoteAttestation(edbHost string) (string, error) {
```

Comment:
shouldn't this live in the edb_atttestation file?

Reply:
edb_attestation.go was entirely copy-pasted and I've been treating it like a third-party lib. That's why it reads so weirdly and I've been using it's API even though it doesn't fit that well. Was hoping we'd be able to find a way to depend on the edgeless ERA tool instead of duplicating it at some point.

But I think from all your comments on it and Joel's I'll just rewrite it, we can own that code as an obscuro implementation of ERA, it'll have just the bits we need and an api that makes sense to us.

Reply:
gotcha